Src/FluentAssertions/Common/MethodInfoExtensions.cs

```diff
@@ -3,27 +3,64 @@
 using System.Linq;
 using System.Linq.Expressions;
 using System.Reflection;
+using System.Runtime.CompilerServices;
 
 namespace FluentAssertions.Common
 {
     public static class MethodInfoExtensions
     {
+        private static Lazy<int> ImplementationOptionsMask = new Lazy<int>(() => Enum.GetValues(typeof(MethodImplOptions)).Cast<int>().Sum(x => x));
+
         internal static bool IsAsync(this MethodInfo methodInfo)
         {
             return methodInfo.GetMatchingAttributes<Attribute>(a => a.GetType().FullName.Equals("System.Runtime.CompilerServices.AsyncStateMachineAttribute")).Any();
         }
 
         internal static IEnumerable<TAttribute> GetMatchingAttributes<TAttribute>(this MemberInfo memberInfo, Expression<Func<TAttribute, bool>> isMatchingAttributePredicate) where TAttribute : Attribute
         {
-            return memberInfo.GetCustomAttributes(
+            var customAttributes = memberInfo.GetCustomAttributes(
                 typeof(TAttribute), false)
                 .Cast<TAttribute>()
+                .ToList();
+
+            if(typeof(TAttribute) == typeof(MethodImplAttribute))
```

Comment:
I would add a check `memberInfo` is an `MethodInfo` here and send the casted value to `RecreateMethodImplAttribute`.

And then tests for `[Not]BeDecoratedWith` when calling on a non-`MethodInfo` member types.

Reply:
Good point, will do that.